public/app/features/dashboard/components/PanelEditor/utils.ts

```diff
@@ -68,6 +68,15 @@ export function setOptionImmutably<T extends object>(options: T, path: string |
   const splat = !Array.isArray(path) ? path.split('.') : path;
 
   const key = splat.shift()!;
+  if (key.endsWith(']')) {
+    let idx = key.lastIndexOf('[');
+    const index = +key.substring(idx + 1, key.length - 1);
+    const propKey = key.substr(0, idx);
+    const current = (options as Record<string, any>)[propKey];
+    const arr = Array.isArray(current) ? [...current] : []; // note spread behavior for missing values is preferable
+    arr[index] = splat.length ? setOptionImmutably(current, splat, value) : value;
+    return { ...options, [propKey]: arr };
+  }
 
   if (!splat.length) {
```

Comment:
I guess this is the break condition for the recursiveness, can it be placed higher up in the function?

Reply:
note that the array case above also has a return.  (with the added behavior of forcing array behavior)